2.4 changelog (#2637)
api-platform/core: CHANGELOG.md

# Changelog

> ## 2.4

Reviewer: ```suggestion
## 2.4.0
```

* Listeners are now opt-in when not handling API Platform operations
* `DISTINCT` is not used when there are no joins

> * Preserve manual join in FilterEagerLoadingExtension #2622

Reviewer 2: ```suggestion
* Preserve manual join in FilterEagerLoadingExtension
```

> * Preserve manual join in FilterEagerLoadingExtension #2622

Reviewer: ```suggestion
* Preserve manual join in `FilterEagerLoadingExtension`
```

> * `elasticsearch` can be disabled resource-wise or per-operation

Reviewer: ```suggestion
* The `elasticsearch` attribute can be disabled resource-wise or per-operation
```

> * The `messenger` attribute can now take the `input` string as a value (`messeger="input"`). This will use a default transformer so that the given `input` is directly sent to the messenger handler.

Reviewer 2: ```suggestion
* The `messenger` attribute can now take the `input` string as a value (`messenger="input"`). This will use a default transformer so that the given `input` is directly sent to the messenger handler.
```

> * The `messenger` attribute can now take the `input` string as a value (`messeger="input"`). This will use a default transformer so that the given `input` is directly sent to the messenger handler.

Reviewer: ```suggestion
* The `messenger` attribute can now take the `input` string as a value (`messenger="input"`). This will use a default transformer so that the given `input` is directly sent to the messenger handler.
```

* The `messenger` attribute can be declared per-operation

> * Mercure updates are published on doctrine flush instead of `kernel.terminate` so that we can couple mercure and messenger

Reviewer: ```suggestion
* Mercure updates are now published after the Doctrine flush event instead of on `kernel.terminate`, so the Mercure and the Messenger integration can be used together
```

> * Use symfony's MetadataAwareNameConverter when available

Reviewer 2: ```suggestion
* Use Symfony's MetadataAwareNameConverter when available
```

> * Use symfony's MetadataAwareNameConverter when available

Reviewer: ```suggestion
* Use Symfony's `MetadataAwareNameConverter` when available
```

> * Change the extension's priorities (`<0`). If you have custom extensions we recommend to use positive priorities.

Reviewer: ```suggestion
* Change the extension's priorities (`<0`) for improved compatibility with Symfony's autoconfiguration feature. If you have custom extensions we recommend to use positive priorities.
```

| Service name                                               | Priority | Class                                              |
|------------------------------------------------------------|------|---------------------------------------------------------|
| api_platform.doctrine.orm.query_extension.eager_loading (collection) | -8 | ApiPlatform\Core\Bridge\Doctrine\Orm\Extension\EagerLoadingExtension |
| api_platform.doctrine.orm.query_extension.eager_loading (item) | -8 | ApiPlatform\Core\Bridge\Doctrine\Orm\Extension\EagerLoadingExtension |
| api_platform.doctrine.orm.query_extension.filter | -16 | ApiPlatform\Core\Bridge\Doctrine\Orm\Extension\FilterExtension |
| api_platform.doctrine.orm.query_extension.filter_eager_loading | -17 | ApiPlatform\Core\Bridge\Doctrine\Orm\Extension\FilterEagerLoadingExtension |
| api_platform.doctrine.orm.query_extension.order | -32 | ApiPlatform\Core\Bridge\Doctrine\Orm\Extension\OrderExtension |
| api_platform.doctrine.orm.query_extension.pagination | -64 | ApiPlatform\Core\Bridge\Doctrine\Orm\Extension\PaginationExtension |

> * Fix JSONLD embeded contexts with outputs

Reviewer: ```suggestion
* Fix JSON-LD contexts when using output classes
```

> * GraphQl: Fix pagination

Reviewer 2: ```suggestion
* GraphQl: Fix pagination (the `endCursor` behavior was wrong)
```

* GraphQl: Improve output/input behavior

> * GraphQl: Improve mutations

Reviewer 2: ```suggestion
* GraphQl: Improve mutations (make the `clientMutationId` nullable and return mutation payload as an object)
```

> * MongoDB: Don't use DbRef for associations

Reviewer 2: ```suggestion
* MongoDB: Fix search filter when searching by related collection id
```

> * MongoDB: Fix numeric and range filters coercion and lte/gte parameters

Reviewer 2: ```suggestion
* MongoDB: Fix numeric and range filters
```
